### PR TITLE
Audit 6 - 2 Step process for critical ops

### PR DESCRIPTION
**Pull request by ShaneDuncan602**
Recommendations
Short term, implement a two-step process for all irrecoverable critical operations. Split all
functions that set contracts owners and administrators into two functions:
1. proposeOwner() or proposeAppManagerAddress() which proposes a new contract
owner or manager which is saved in the contract storage,
2. confirmOwner() or confirmAppManagerAddress() which can only be called by the
proposed owner or manager and is used to finalize the transfer

This will guarantee that the owner-setting party must be able to call the contract from the
address in question before it is actually set as the new owner.

### CODE REVIEW

File: src/application/IAppManagerUser.sol [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * @title App Manager User Interface
+ * @dev This interface is implemented by all contracts that use AppManager. It provides the common function for setting up a new link to an AppManager
+ * @author @ShaneDuncan602, @oscarsernarosero, @TJ-Everett
+ * @notice Interface for app manager user functions.
+ */
+interface IAppManagerUser {
+    /**
+     * @dev this function confirms a new appManagerAddress that was put in storageIt can only be confirmed by the proposed address
```
Comment (TJ-Everett) [inline src/application/IAppManagerUser.sol]: This* and missing a . after storage

**Review by TJ-Everett — APPROVED** [review]
Great work Shane this was high effort for sure. Just the one little typo. Approving

File: src/data/DataModule.sol [hunk 2]
```diff
@@ -27,19 +30,28 @@ contract DataModule is IDataModule, Ownable {
...
     /**
-     * @dev Transfers ownership of the contract to a new account (`newOwner`).
-     * Can only be called by the current owner.
+     * @dev this function confirms a new appManagerAddress that was put in storage. It can only be confirmed by the proposed address
      */
-    function transferDataOwnership(address newOwner) public appAdminstratorOrOwnerOnly {
-        transferOwnership(newOwner);
+    function confirmOwner() external {
+        if (newOwner == address(0)) revert NoProposalHasBeenMade();
+        if (msg.sender != newOwner) revert ConfirmerDoesNotMatchProposedAddress();
```
Comment (oscarsernarosero) [inline src/data/DataModule.sol:46]: Interesting. So the confirmation has to be made from the new owner? kinda like accepting the ownership?

Comment (oscarsernarosero) [reply src/data/DataModule.sol:46]: I love it Shane

File: src/data/DataModule.sol [hunk 3]
```diff
@@ -27,19 +30,28 @@ contract DataModule is IDataModule, Ownable {
...
+        _transferOwnership(newOwner);
+    }
+
+    /**
+     * @dev Part of the two step process to set a new Data Provider within a Protocol AppManager
+     * @param _providerType the type of data provider
+     */
+    function confirmDataProvider(ProviderType _providerType) external virtual appAdminstratorOrOwnerOnly {
```
Comment (oscarsernarosero) [inline src/data/DataModule.sol:54]: Shouldn't we re-enter the address here to make sure we proposed the right address in the first step?

File: src/application/AppManager.sol [hunk 4]
```diff
@@ -812,47 +818,87 @@ contract AppManager is IAppManager, AccessControlEnumerable, IAppLevelEvents {
         appName = _appName;
     }
 
+    /**
+     * @dev Part of the two step process to set a new AppManager within a Protocol Entity
+     * @param _assetAddress address of a protocol entity that uses AppManager
+     */
+    function confirmAppManager(address _assetAddress) external onlyAppAdministrator {
+        IAppManagerUser(_assetAddress).confirmAppManagerAddress();
+    }
```
Comment (oscarsernarosero) [inline src/application/AppManager.sol:827]: Super nice

**Review by oscarsernarosero — APPROVED** [review]
Shane this looks really really amazing. Fantastic job

File: src/data/DataModule.sol (hunk 3 above)
Comment (ShaneDuncan602) [reply src/data/DataModule.sol:54]: It has a check in the appManager to make sure the sender is the new proposed data provider.

**Comment by ShaneDuncan602** [comment]
merged with external

**Pull request closed without merging**